packages/utilities/src/array.ts

```diff
@@ -78,4 +78,46 @@ export function toMatrix<T>(items: T[], columnCount: number): T[][] {
  */
 export function removeIndex<T>(array: T[], index: number): T[] {
   return array.filter((_: T, i: number) => index !== i);
+}
+
+/**
+ * Takes an array, creates a copy, applies a mutation function on the copy, and returns the modified copy.
+ * This can be used to transform a function that mutates an array into one that returns a modified copy instead.
+ * @param array - Array to me manipulated
+ * @param mutateFunction - The function that will mutate the copied array
+ */
+function copyAndMutateArray<T>(array: T[], mutateFunction: (arr: T[]) => void): T[] {
+  const copy = array.slice();
+  mutateFunction(copy);
+  return copy;
+}
+
+/**
+ * Given an array, this function returns a new array where the element at a given index has been replaced.
+ * @param array - The array to operate on
+ * @param newElement - The element that will be placed in the new array
+ * @param index - The index of the element that should be replaced
+ */
+export function replaceElement<T>(array: T[], newElement: T, index: number): T[] {
+  return copyAndMutateArray(array, (arr: T[]): void => { arr[index] = newElement; });
+}
+
+/**
+ * Given an array, this function returns a new array where an element has been inserted at the given index.
+ * @param array - The array to operate on
+ * @param index - The index where an element should be inserted
+ * @param itemToAdd - The element to insert
+ */
+export function addElementAtIndex<T>(array: T[], index: number, itemToAdd: T): T[] {
+  return copyAndMutateArray(array, (arr: T[]): void => { arr.splice(index, 0, itemToAdd); });
```

Comment:
Maybe I should get rid of the copy and mutate to avoid the arrow function overhead on each call.